File: DataStructure/binary_search.cpp
#include<iostream>
#include<vector>

using namespace std;

/**
 * 二分查找和边界查找
 *
*/
int find(const vector<int>& arr, int tgt){
	int left = 0;
	int right = arr.size() - 1;

	while (left <= right){
		int mid = left + (right - left) / 2;
		if (arr[mid] == tgt){
			return mid;
		}
		if (arr[mid] > tgt){
			right = mid - 1;
		}else if (arr[mid] < tgt){
			left = mid + 1;
		}
	}
	return -1;
}

int find_left(vector<int>& arr, int tgt){
	int left = 0;
	int right = arr.size() - 1;

	while (left <= right){
		int mid = left + (right - left) / 2;
		if (arr[mid] == tgt){
			right = mid - 1;
		}
		if (arr[mid] > tgt){
			right = mid - 1;
		}
		else if (arr[mid] < tgt){
			left = mid + 1;
		}
	}
	return left;
}

int find_right(vector<int>& arr, int tgt){
	int left = 0;
	int right = arr.size() - 1;

	while (left <= right){
		int mid = left + (right - left) / 2;
		if (arr.at(mid) == tgt){
			left = mid + 1;
		}else if (arr.at(mid) > tgt){
			right = mid - 1;
		}else if (arr.at(mid) < tgt){
			left = mid + 1;
		}
	}
	return right;
}

int main(){
	vector<int> arr = {1, 1, 2, 2, 2, 5, 6, 7};

	cout << find_left(arr, 2) << endl;
	cout << find_right(arr, 2) << endl;

	return 0;
}

File: DataStructure/binary_search.py
# -*- coding: utf-8 -*-
# Id$
# @file
# @author zhujisheng@corp.netease.com(Zoo)

def find(arr, tgt):
	i = 0
	j = len(arr)

	while i < j:
		mid = i + (j - i) // 2
		if arr[mid] == tgt:
			return mid
		if arr[mid] < tgt:
			i = mid + 1
		else:
			j = mid - 1
	return -1

def find_left(arr, tgt):
	i = 0
	j = len(arr)

	while i < j:
		mid = i + (j - i) // 2
		if arr[mid] == tgt:
			j = mid - 1
		elif arr[mid] > tgt:
			j = mid - 1
		else:
			i = mid + 1
	if i < len(arr) and arr[i] == tgt:
		return i
	return -1

def find_right(arr, tgt):
	i = 0
	j = len(arr)

	while i < j:
		m = i + (j - i) // 2
		if arr[m] == tgt:
			i = m + 1
		elif arr[m] < tgt:
			i = m + 1
		else:
			j = m - 1
	if 0 <= j < len(arr) and arr[j] == tgt:
		return j
	return -1

if __name__ == "__main__":
	arr = range(1, 100)
	arr = [1, 2, 3, 3, 3, 3, 4, 5]
	print find_right(arr, 3)

